xclim/sdba/base.py

```diff
@@ -16,6 +18,64 @@ class Parametrizable(dict):
     :py:meth:`Parametrizable.parameters` dictionary, the copy method and the class representation.
     """
 
+    _parametrizable_classes = {}
+
+    def __new__(cls, *args, **kwargs):
+        """Add subclass to registry."""
+        cls._parametrizable_classes[cls.__name__] = cls
+        return super().__new__(cls, *args, **kwargs)
+
+    @classmethod
+    def from_string(cls, defstr: str):
+        """Create an instance of this object from a string.
+
+        The string must have been created with `str(object)` where object has the same type
+        as this object. The resulting instance might be slightly different than the original one.
+
+        Parameters
+        ----------
+        defstr : str
+          String providing the definition of the object.
+
+        Examples
+        --------
+        >>> obj = Parametrizable(anint=1, anobj=Parametrizable(astr='hello world'))
+        >>> ss = str(obj)
+        >>> obj2 = Parametrizable.from_string(ss)
+        >>> obj
+        Parametrizable(anint=1, anobj=Parametrizable(astr='hello world'))
+        >>> obj2
+        Parametrizable(anint=1, anobj=Parametrizable(astr='hello world'))
+        """
+        # regex pattern that fully matches a string produces by Parametrizable._repr__
+        match = re.fullmatch(r"(?P<class>[A-Za-z0-9_]+)\((?P<args>.*)\)", defstr)
+        if not match:
+            raise ValueError(
+                f"Received malformed string for creating a {cls.__name__} instance. ({defstr})"
+            )
+        if match.groupdict()["class"] != cls.__name__:
+            raise ValueError(
+                f"Wrong class : constructor of {cls.__name__} received a string for {match.groupdict()['class']}"
+            )
+
+        obj = cls.__new__(cls)
+        params = {}
+        # regex pattern that matches each argument in the repr of a Parametrizable obj, distinguishing other Parametrizable classes (and their arguments) from python literals
+        print(match.groupdict()["args"] + ", ")
+        for argmatch in re.finditer(
+            r"(?P<name>[A-Za-z0-9_]+)=(?P<vstr>((?P<vcls>[A-Za-z0-9_]+)\((.*?)\))|.+?(?=, ))",
+            match.groupdict()["args"] + ", ",
+        ):
+            dd = argmatch.groupdict()
+            if dd["vcls"] is not None:
+                params[dd["name"]] = cls._parametrizable_classes[
+                    dd["vcls"]
+                ].from_string(dd["vstr"])
+            else:
+                params[dd["name"]] = literal_eval(dd["vstr"])
+        obj.update(params)
+        return obj
```

Comment:
Not sure I understand what you're trying to do here... could you add some comments ?

Maybe this is unrelated, but if you're trying to cast keyword arguments to a class, you could look into pydantic. It can be used to automatically parse objects and cast them to the class implementing a data model compatible with the inputs.

Reply:
I'll add comments.
Overall this parses the string that it generates in the `__repr__`.  The most difficult thing was to split the string between the arguments and distinguish between normal python literals and our custom "Parametrizable" classes. Thus the regex.
I'm not sure how pydantic could help us here? Does it have string parsing mechanisms?

Reply:
More generally, we are looking for human-readable object serialization. Usual serialization is covered by pickle, but that would create unreadable strings. JSON is an option, but again that string would complex. There is [PyON](https://github.com/lagmoellertim/pyon) that can handle the custom classes in JSON, but it is not very active and not on conda.

Seeing that we do not need to cover many cases and can completely control the custom classes that we need to parse, I think having our own in-house method is ok, and secure enough.

But maybe @Zeitsperre, @tlvu or @cjauvin have ideas about this problem? TL;DR: We want to serialize a python object to a human-readable string and be able to reconstruct it. For now, object are made of simple python literals OR other instances of the same base class, thus sharing the (de-)serialization mechanisms.

However, we could split this in 2 things : normal human-readable repr + string serialization. The "problem" here is simply that it would mean adding a completely unreadable attribute to the datasets we make, an attribute that does not make sense outside of xclim (while the current kinda does).

Reply:
@aulemahal creating a JSON file with both a human-readable portion and a pickle bytestring makes sense to me, honestly. My only concern would be ensuring that some sort of safety mechanism is in place to evaluate that the human-readable portion of this object/file corresponds with the pickled string values and maybe exit early if there's an inconsistency? My biggest worry would be that they don't agree and the values are wrong (beyond all the other concerns from pickles).

Reply:
True that is also a concern : if the human readable attribute and the machine-readable could differ. But to check that goes back to parsing the human readable one! Also, just to precise, we want to store that information in the attribute of a netCDF. So, while a json file is readable, a json _string_ scores lower.

Reply:
This is probably a terrible suggestion, but if we were outputting `zarr` data, this would be really simple. https://zarr.readthedocs.io/en/stable/spec/v2.html#attributes. Again, not saying we do throw the baby out with the bathwater, but this would solve the JSON problem.